substratevm/src/com.oracle.svm.configure/src/com/oracle/svm/configure/config/ConfigurationDynamicClass.java

```diff
@@ -0,0 +1,103 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.configure.config;
+
+import com.oracle.svm.configure.json.JsonPrintable;
+import com.oracle.svm.configure.json.JsonWriter;
+import com.oracle.svm.core.configure.ConfigurationFiles;
+
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public final class ConfigurationDynamicClass implements JsonPrintable {
+
+    private final String qualifiedJavaName;
+    private final String checksum;
+    private final String classFile;
+    private final byte[] classContents;
+
+    public static ConfigurationDynamicClass newTraceTimeConfig(String qualifiedJavaName, String checksum, byte[] classContents) {
+        return new ConfigurationDynamicClass(qualifiedJavaName, checksum, null, classContents);
+    }
+
+    public static ConfigurationDynamicClass newBuildTImeConfig(String qualifiedJavaName, String checksum, String classFile) {
+        return new ConfigurationDynamicClass(qualifiedJavaName, checksum, classFile, null);
+    }
+
+    private ConfigurationDynamicClass(String qualifiedJavaName, String checksum, String classFile, byte[] classContents) {
+        assert qualifiedJavaName.indexOf('/') == -1 : "Requires qualified Java name, not internal representation";
+        assert !qualifiedJavaName.startsWith("[") : "Requires Java source array syntax, for example java.lang.String[]";
+        this.qualifiedJavaName = qualifiedJavaName;
+        this.checksum = checksum;
+        this.classContents = classContents;
+        this.classFile = classFile;
+    }
+
+    public boolean contains(String definedClassName, String checksum2Match) {
+        return definedClassName != null && definedClassName.equals(qualifiedJavaName) && checksum2Match.equals(this.checksum);
+    }
+
+    @Override
+    public void printJson(JsonWriter writer) throws IOException {
+        // Mangle class' qualified name into a system independent format
+        String internalName = qualifiedJavaName.replace(".", ConfigurationFiles.MANGLE_SLASH);
```

Comment:
As now I directly read the class file into byte array to call `defineClass` instead of finding class from class path, there is no need to keep the package information in the file name. Actually the file name can be anything that is unique.
What do you think using the already calculated SHA value as file name? In this way the file name is unique, and furthermore, we can save one entry in configuration file. As the file name is from SHA, only one of `classFile` and `checksum` entry is needed.

Reply:
Yes, I believe the way forward is using a unique hash code in the file name, potentially the unqualified name too for identifiability to the developer.